lsm/structs.go

```diff
@@ -0,0 +1,639 @@
+	buff = encoding.AppendUint64ToBufferLE(buff, uint64(l.TotTables))
+	buff = encoding.AppendUint64ToBufferLE(buff, uint64(l.BytesIn))
+	buff = encoding.AppendUint64ToBufferLE(buff, uint64(l.EntriesIn))
+	buff = encoding.AppendUint64ToBufferLE(buff, uint64(l.TablesIn))
+	buff = encoding.AppendUint64ToBufferLE(buff, uint64(l.TotCompactions))
+	buff = encoding.AppendUint32ToBufferLE(buff, uint32(len(l.LevelStats)))
+	for level, levStats := range l.LevelStats {
+		buff = encoding.AppendUint32ToBufferLE(buff, uint32(level))
+		buff = levStats.serialize(buff)
+	}
+	return buff
+}
+
+func (l *Stats) Deserialize(buff []byte, offset int) int {
+	var u uint64
+	u, offset = encoding.ReadUint64FromBufferLE(buff, offset)
+	l.TotBytes = int(u)
+	u, offset = encoding.ReadUint64FromBufferLE(buff, offset)
+	l.TotEntries = int(u)
+	u, offset = encoding.ReadUint64FromBufferLE(buff, offset)
+	l.TotTables = int(u)
+	u, offset = encoding.ReadUint64FromBufferLE(buff, offset)
+	l.BytesIn = int(u)
+	u, offset = encoding.ReadUint64FromBufferLE(buff, offset)
+	l.EntriesIn = int(u)
+	u, offset = encoding.ReadUint64FromBufferLE(buff, offset)
+	l.TablesIn = int(u)
+	u, offset = encoding.ReadUint64FromBufferLE(buff, offset)
+	l.TotCompactions = int(u)
+	var nl uint32
+	nl, offset = encoding.ReadUint32FromBufferLE(buff, offset)
+	numLevels := int(nl)
+	levStatsMap := make(map[int]*LevelStats, numLevels)
+	for i := 0; i < numLevels; i++ {
+		var l uint32
+		l, offset = encoding.ReadUint32FromBufferLE(buff, offset)
+		levStats := &LevelStats{}
+		offset = levStats.deserialize(buff, offset)
+		levStatsMap[int(l)] = levStats
+	}
+	l.LevelStats = levStatsMap
+	return offset
+}
+
+func (l *Stats) copy() *Stats {
+	statsCopy := &Stats{}
+	statsCopy.TotBytes = l.TotBytes
+	statsCopy.TotEntries = l.TotEntries
+	statsCopy.TotTables = l.TotTables
+	statsCopy.BytesIn = l.BytesIn
+	statsCopy.EntriesIn = l.EntriesIn
+	statsCopy.TablesIn = l.TablesIn
+	statsCopy.TotCompactions = l.TotCompactions
+	statsCopy.LevelStats = make(map[int]*LevelStats, len(l.LevelStats))
+	for level, levStats := range l.LevelStats {
+		statsCopy.LevelStats[level] = &LevelStats{
+			Bytes:   levStats.Bytes,
+			Entries: levStats.Entries,
+			Tables:  levStats.Tables,
+		}
+	}
+	return statsCopy
+}
+
+type LevelStats struct {
+	Bytes   int
+	Entries int
+	Tables  int
+}
+
+func (l *LevelStats) serialize(buff []byte) []byte {
+	buff = encoding.AppendUint64ToBufferLE(buff, uint64(l.Bytes))
+	buff = encoding.AppendUint64ToBufferLE(buff, uint64(l.Entries))
+	buff = encoding.AppendUint64ToBufferLE(buff, uint64(l.Tables))
+	return buff
+}
+
+func (l *LevelStats) deserialize(buff []byte, offset int) int {
+	var u uint64
+	u, offset = encoding.ReadUint64FromBufferLE(buff, offset)
+	l.Bytes = int(u)
+	u, offset = encoding.ReadUint64FromBufferLE(buff, offset)
+	l.Entries = int(u)
+	u, offset = encoding.ReadUint64FromBufferLE(buff, offset)
+	l.Tables = int(u)
+	return offset
+}
+
+type MasterRecord struct {
```

Comment:
to confirm my base level architecture understanding - this is the same for every agent in the distributed system, right? Its source of truth is S3. (this is how I see the old `levels/structs.go` master record get created

Reply:
There would be one master record for each instance of the LSM, and then we shard the overall database by having multiple LSM instances, each covering different key ranges. These instances live on different agents in the cluster to distribute the load more evenly. This is the source of truth for the LSM instance and records which SSTables are in which levels of the LSM (amongst other things), so it holds the structure of the LSM.